cycles-ledger/src/storage.rs

```diff
@@ -1567,25 +1672,55 @@ pub async fn withdraw(
         // subtract the fee to pay for the reimburse block
         let amount_to_reimburse = amount.saturating_sub(config::FEE);
         if amount_to_reimburse.is_zero() {
-            return Err(FailedToWithdraw {
+            return Err(FailedToWithdrawFrom {
                 fee_block: None,
+                approval_refund_block: None,
                 rejection_code,
                 rejection_reason,
             });
         }
         match reimburse(from, amount_to_reimburse, now) {
             Ok(fee_block) => {
                 prune(now);
-                return Err(FailedToWithdraw {
+                if let Some(spender) = spender {
+                    if spender != from && amount > 3 * config::FEE {
+                        match reimburse_approval(
+                            from,
+                            spender,
+                            amount_with_fee.saturating_sub(3 * config::FEE), // charge FEE for every block: withdraw attempt, refund, refund approval
+                            old_expires_at,
```

Comment:
I don't think we should pass the `old_expires_at`. `expires_at` should not be touched by the `reimburse_approval` function. The resolution between the parameter and the state of the Ledger is confusing and I'm not sure what the advantage would be.

Reply:
If the initial `withdraw` used the entire allowance then the allowance entry is deleted. As a result we don't have any memory of when the approval would have expired. If we then need to reimburse the allowance, what `expires_at` would you use besides `old_expires_at`?

Reply:
That makes a lot of sense. The expiry time must be captured in a variable as otherwise an approval cannot be reinstated.